style: apply /style-guide pass to models/integrations (#2673)
wandb/docs: models/integrations/add-wandb-to-any-library.mdx

---
title: Add W&B to a Python library
description: Best practices for integrating W&B into your Python library for experiment tracking, system monitoring, and model management.
keywords: ["library integration", "WandbCallback pattern", "publish integration"]
---

import ApiKeyCreateStreamlined from "/snippets/_includes/api-key-create-streamlined.mdx";

This guide explains how to integrate W&B into a Python library so that your users can track experiments, monitor system metrics, and manage models when they use your code. It's intended for library authors and maintainers who want to expose W&B functionality through their own framework, SDK, or reusable training code.

Follow these recommendations if you're integrating W&B into a complex codebase (such as a training framework, SDK, or reusable library) where the codebase is more involved than a single Python training script or Jupyter notebook.

<Tip>
If you're new to W&B, review the core guides (for example, [Experiment Tracking](/models/track/)) before continuing.
</Tip>

> The following sections walk through the major integration decisions in order: how users install W&B, how they authenticate, how to start and configure runs, how to log metrics and artifacts, and how to support distributed training and hyperparameter sweeps.

Reviewer: ```suggestion
The following sections walk through the major integration decisions in order: how to install W&B, how to authenticate, how to start and configure runs, how to log metrics and artifacts, and how to support distributed training and hyperparameter sweeps.
```

Reviewer: Why in third person?

Author: Reworded to second person ('how to install W&B, how to authenticate…') per the suggestion.

## Decide how users install W&B

> Before you start, decide whether W&B should be a required dependency or an optional feature of your library. This choice affects how you import `wandb`, how you document installation, and how you handle environments where `wandb` isn't present.

Reviewer: Would prefer if, on the first mention, call out that `wandb` is the W&B Python SDK.

Author: Done — the first mention now reads 'the W&B Python SDK (`wandb`)'.

### Install and log in to W&B

After you have an API key, install the `wandb` library locally and log in so that subsequent runs can authenticate to W&B. Choose the tab that matches your environment.

<Tabs>
<Tab title="Command Line">
1. Set the `WANDB_API_KEY` [environment variable](/models/track/environment-variables/) to your API key:

    ```bash

>     export WANDB_API_KEY=[YOUR-API-KEY]

Reviewer: For consistency, remove brackets. (Elsewhere in the docs we use open and close < and >).  We also add a one liner like: 

"Replace values enclosed in `<>` with your own."

Author: Reverted to `<your_api_key>` and added the lead-in 'Replace values enclosed in `<>` with your own:'; propagated to every login block.

## Start a run

After you set up authentication, the next step is to start a W&B run so that your library has somewhere to log metrics, configs, and artifacts.

A *run* represents a single unit of computation, such as a training experiment. Most libraries create one run per training job. For more information about runs, see [W&B Runs](/models/runs/).

Initialize a run with [`wandb.init()`](/models/ref/python/functions/init) and specify a name for your project and your team entity (team name). If you don't specify a project, W&B stores your run in a default project called "uncategorized":

```python

> with wandb.init(project="[PROJECT-NAME]", entity="[ENTITY]") as run:

Reviewer: See previous comment about open and close "<>".

Author: Done — now `<project_name>` / `<entity>`.

    <img src="/images/integrations/integrations_add_any_lib_log.png" alt="W&B Workspace"  />
</Frame>

For more information, see [`wandb.Run.log()`](/models/ref/python/experiments/run#method-run-log).

### Control the x-axis

> By default, the `wandb` SDK manages its own step counter, which might not match the step semantics of your training loop. If you perform multiple calls to `wandb.Run.log()` for the same training step, the `wandb` SDK increments an internal step counter for each call to `wandb.Run.log()`. This counter might not align with the training step in your training loop.

Reviewer: ```suggestion
By default, the W&B Python SDK manages its own step counter, which might not match the step semantics of your training loop. If you perform multiple calls to `wandb.Run.log()` for the same training step, the `wandb` SDK increments an internal step counter for each call to `wandb.Run.log()`. This counter might not align with the training step in your training loop.
```

Author: Applied.

    run.log({"global_step": step, "eval/loss": 0.2})
```

> If you don't have access to the independent step variable (for example, `global_step` isn't available during your validation loop), `wandb` automatically uses the previously logged value for `global_step`. In this case, ensure you log an initial value for the metric so that it's defined when it's needed.

Reviewer: ```suggestion
If you don't have access to the independent step variable (for example, `global_step` isn't available during your validation loop), W&B automatically uses the previously logged value for `global_step`. In this case, ensure you log an initial value for the metric so that it's defined when it's needed.
```

Reviewer: Refer to W&B not `wandb` as the entity making the action.

Author: Applied — W&B (not `wandb`) is now the actor.

run.log_artifact(artifact, aliases=aliases)
```

> The preceding snippet logs a model checkpoint as an artifact with metadata such as evaluation accuracy and training steps. The artifact's name includes the unique run ID, and it's tagged with [custom aliases](/models/artifacts/create-a-custom-alias/) for quick reference.

Reviewer: ```suggestion
The previous snippet logs a model checkpoint as an artifact with metadata such as evaluation accuracy and training steps. The artifact's name includes the unique run ID, and it's tagged with [custom aliases](/models/artifacts/create-a-custom-alias/) for quick reference.
```

We no longer user "preceding" and instead use "previous".

Author: Applied; also swept 'preceding'→'previous' across the set.

### Log input artifacts

To capture lineage between data and models, log the datasets or pretrained models that a run consumes as inputs:

```python
dataset = wandb.Artifact(name="flowers", type="dataset")
dataset.add_file("flowers.npy")
run.use_artifact(dataset)
```

> The preceding snippet creates an artifact for a dataset called "flowers" and adds a file to it. The `run.use_artifact()` call associates the artifact with the current run so that W&B can track the lineage of the dataset used in the run.

Reviewer: ```suggestion
The previous snippet creates an artifact for a dataset called "flowers" and adds a file to it. The `wandb.Run.use_artifact()` call associates the artifact with the current run so that W&B can track the lineage of the dataset used in the run.
```

Author: Applied — `wandb.Run.use_artifact()`.

### Download artifacts

> After you log artifacts, your library (or its users) can download previously logged artifacts from W&B to use in training or inference code. The right approach depends on whether you already have an active run.

Reviewer: What does "(or its users)" mean in this sentence?

Author: Removed the parenthetical — reworded to 'After you log artifacts, you can download…'.

> If you have a run context, use [`wandb.Run.use_artifact()`](/models/ref/python/experiments/run) to reference an artifact in W&B and then call [`wandb.Artifact.download()`](/models/ref/python/experiments/artifact) to download it to a local directory. Using `use_artifact()` also records the artifact as an input to the current run, preserving lineage.

Reviewer: ```suggestion
If you have a run context, use [`wandb.Run.use_artifact()`](/models/ref/python/experiments/run) to reference an artifact in W&B and then call [`wandb.Artifact.download()`](/models/ref/python/experiments/artifact) to download it to a local directory. Using `wandb.Run.use_artifact()` also records the artifact as an input to the current run, preserving lineage.
```

Author: Applied — `wandb.Run.use_artifact()`.

```python
with wandb.init(...) as run:
    artifact = run.use_artifact("user/project/artifact:latest")
    local_path = artifact.download()
```

Use the [W&B Public API](/models/ref/python/public-api/) to reference and download an artifact without initializing a run. This is useful in scenarios such as distributed environments or when you perform inference, where you might not want to create a new run.

```python
import wandb
artifact = wandb.Api().artifact("user/project/artifact:latest")
local_path = artifact.download()
```

For more information, see [Download and use artifacts](/models/artifacts/download-and-use-an-artifact/).

## Tune hyperparameters

> If your library supports hyperparameter tuning, you can integrate [W&B Sweeps](/models/sweeps/) to manage and visualize experiments. Sweeps coordinate multiple runs across a defined search space and surface the results in the W&B App so users can compare configurations side by side.

Reviewer: Did Matt say something about no longer refer to the W&B App as "W&B App"? I think it has a formal name now. Not sure.

Author: Removed all 'W&B App' references; the web app is just 'W&B' now.
